[#9] [FEAT] 프로젝트 패키지 및 엔티티 맵핑

jiss02 opened the pull request:
도메인 별 패키지를 생성하고 엔티티를 맵핑하였습니다.

resolve #9

jiss02 committed:
[#9] [FEAT] 프로젝트 패키지 및 엔티티 맵핑

도메인 별 패키지를 생성하고 엔티티를 맵핑하였습니다.

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/approval/domain/Approval.java:
```diff
@@ -0,0 +1,43 @@
+
+@Getter
+@Entity
+@NoArgsConstructor
```
현재 코드를 보면 Approval이라는 엔티티의 객체 생성 할수 있는 방법은 디폴트 생성자와 빌더패턴을 이용한 객체 생성 두가지가 있는데요. 빌더패턴을 사용하면 디폴트생성자로 객체 생성할 이유가 있을까요??

jiss02 replied:
디폴트 생성자를를 붙이라는 오류 메세지가 계속 떠서 적은 어노테이션인데, 방법을 더 찾아보겠습니다.

korea8378 replied:
접근지정자를 이용해보세요. public, private, protected, default 중에 머가 좋을까요?

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/approval/domain/Approval.java:
```diff
@@ -0,0 +1,43 @@
+    @ManyToOne
+    @JoinColumn(name = "user_id")
+    @MapsId("userId")
+    private User user;
```
ApprovalId라는 복합키를 사용하고 있는중인데 User라는 맵핑 관계가 필요할까요???

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/approval/domain/Approval.java:
```diff
@@ -0,0 +1,43 @@
+    @ManyToOne
+    @JoinColumn(name = "doc_id")
+    @MapsId("docId")
+    private Document document;
```
ApprovalId라는 복합키를 사용하고 있는중인데 Document라는 맵핑 관계가 필요할까요???(위와 동일한 이유입니다.)

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/approval/domain/Approval.java:
```diff
@@ -0,0 +1,43 @@
+    private Integer ordering;
+
+    @Column(length = 6)
+    private String state;
```
state라는 필드만 보고는 어떠한 상태들이 있는지 파악하기가 어려울꺼 같습니다. 좀 더 효율적으로 상태를 관리할수 있는 방법이 있지않을가요??

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/approval/domain/ApprovalId.java:
```diff
@@ -0,0 +1,19 @@
+import javax.persistence.*;
+import java.io.Serializable;
+
+@Data
```
Lombok의 @data 어노테이션을 사용하면 getter, setter가 만들어지는데요. @Data보다는 정말로 필요한 어노테이션 추가해서 메서드를 제네레이션하는게 좋을것 같습니다.

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/approval/controller/SignController.java:
```diff
@@ -0,0 +1,9 @@
+
+@RequiredArgsConstructor
+@RestController
+public class SignController {
```
도메인을 Approval로 가져가면 컨트롤러 네이밍을 Approval으로 맞쳐주는게 좋을것 같습니다.

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/approval/service/SignService.java:
```diff
@@ -0,0 +1,9 @@
+
+@RequiredArgsConstructor
+@Service
+public class SignService {
```
도메인을 Approval로 가져가면 서비스 네이밍을 Approval으로 맞쳐주는게 좋을것 같습니다.

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/document/domain/Document.java:
```diff
@@ -0,0 +1,39 @@
+
+@Getter
+@Entity
+@NoArgsConstructor
```
현재 코드를 보면 Document이라는 엔티티의 객체 생성 할수 있는 방법은 디폴트 생성자와 빌더패턴을 이용한 객체 생성 두가지가 있는데요. 빌더패턴을 사용하면 디폴트생성자로 객체 생성할 이유가 있을까요??

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/document/domain/Document.java:
```diff
@@ -0,0 +1,39 @@
+    private String title;
+
+    @Column(length=6)
+    private String state;
```
state라는 필드만 보고는 어떠한 상태들이 있는지 파악하기가 어려울꺼 같습니다. 좀 더 효율적으로 상태를 관리할수 있는 방법이 있지않을가요??

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/document/domain/Document.java:
```diff
@@ -0,0 +1,39 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name="user_id", nullable = false)
+    private User user;
```
문서 엔티티가 유저 엔티티를 직접 가지고 있을만큼 강한 의존성이 필요할까요??? 유저라는 도메인이 인증프로젝트를 위해 별도로 분리된다면 유저라는 엔티티(객체)를 참조할수있을까요??

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/document/domain/Document.java:
```diff
@@ -0,0 +1,39 @@
+    private User user;
+
+    @Builder
+    public Document(User user, String content, String title, String state){
```
빌더패턴을 사용한다면 다른생성자는 private을 통하여 빌더패턴만 사용하게 해주는게 좋습니다.

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/user/domain/RoleType.java:
```diff
@@ -0,0 +1,13 @@
+import lombok.RequiredArgsConstructor;
+
+@Getter
+@RequiredArgsConstructor
```
enum은 보통 객체를 이미 생성하여 상수로 사용하기 위해 사용이됩니다. @RequiredArgsConstructor만 사용한다면 접근지정자가 기본적으로 public인데요. 그렇다면 다른 외부에서 RoleType이라는 객체를 생성할수 있지 않을까요?

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/user/domain/User.java:
```diff
@@ -0,0 +1,40 @@
+    private String rank;
+
+    @Builder
+    public User(RoleType roleType, String password, String name, String rank) {
```
빌더패턴을 사용한다면 다른생성자는 private을 통하여 빌더패턴만 사용하게 해주는게 좋습니다.

jiss02 committed:
[#9] [REFACTOR] 도메인 구조 변경 및 엔티티 매핑 수정

- 기존의 document와 approval을 합쳐 도메인 구조를 변경하였습니다.
- 도메인 이름과 contoller의 이름을 일치시켰습니다.
- 엔티티 매핑에서의 연관 관계를 해제하였습니다.
- 엔티티 생성자에 대한 접근 제어자를 변경했습니다.
- 문서의 state필드에 대한 enum을 생성했습니다.
- Builder패턴에서 private가 아닌 부분들을 수정하였습니다.
- 복합키에서의 @Data 어노테이션을 수정하였습니다.

korea8378 commented on src/main/java/mashup/backend/hoowook_hoowwok/document/domain/approval/Approval.java:
```diff
@@ -0,0 +1,33 @@
+
+    @Enumerated(EnumType.STRING)
+    private State state;
+
```
결재 승인/거절시 코멘트를 달수 있도록 했던것 같은데 안보이는거 같네요 추가해주시면 좋을것 같습니다.

jiss02 committed:
[#9] [REFACTOR] approval 엔티티 코멘트 컬럼 추가

- approval에 코멘트 컬럼을 추가하였습니다.
- document의 content 컬럼에 달리는 어노테이션을 @Lob으로 수정하였습니다.

secureWKkim commented:
👍🏻
